--- /dev/null
+++ b/board/olimex/common/boards.h
@@ -0,0 +1,51 @@
+/*
+ * SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+ *
+ * List of all supported devices
+ *
+ * Copyright (C) 2019 Olimex Ltd.
+ *   Author: Stefan Mavrodiev <stefan@olimex.com>
+ */
+#ifndef __BOARDS_H
+#define __BOARDS_H
+
+#include <linux/sizes.h>
+
+enum olinuxino_storage {
+	STORAGE_NONE = 0,
+	STORAGE_EMMC = 'e',
+	STORAGE_NAND = 'n',
+	STORAGE_SPI = 's'
+};
+
+enum olinuxino_grade {
+	GRADE_COM = 0,
+	GRADE_IND = 1,
+};
+
+
+enum olinuxino_sizes {
+	S_1 = 0, S_2, S_4, S_8, S_16, S_32, S_64, S_128, S_256, S_512
+};
+
+#define BYTES(a)	( S_##a )
+#define KBYTES(a)	( S_##a + 10)
+#define MBYTES(a)	( S_##a + 20)
+#define GBYTES(a)	( S_##a + 30)
+
+#define OLINUXINO_CONFIG(__storage, __size, __ram, __grade) \
+	.config = { STORAGE_##__storage, __size, __ram, GRADE_##__grade },
+
+#define OLINUXINO_BOARD(__id, __name) \
+	.id = __id, \
+	.name = __name,
+
+struct olinuxino_boards {
+	uint32_t id;
+	const char name[32];
+	uint8_t config[4];
+};
+
+extern struct olinuxino_boards olinuxino_boards[];
+
+#endif /* __BOARDS_H */
--- /dev/null
+++ b/board/olimex/common/boards.c
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 2019 Olimex Ltd.
+ *   Author: Stefan Mavrodiev <stefan@olimex.com>
+ *
+ * SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+ */
+#include <common.h>
+#include "boards.h"
+
+struct olinuxino_boards olinuxino_boards[] = {
+#if defined(CONFIG_TARGET_A64_OLINUXINO)
+	{
+		OLINUXINO_BOARD(8861, "A64-OLinuXino-2Ge8G-IND")
+		OLINUXINO_CONFIG(EMMC, GBYTES(8), GBYTES(2), IND)
+	},
+	{
+		OLINUXINO_BOARD(9065, "A64-OLinuXino-1Gs16M")
+		OLINUXINO_CONFIG(SPI, MBYTES(16), GBYTES(1), COM)
+	},
+	{
+		OLINUXINO_BOARD(8367, "A64-OLinuXino-1Ge4GW")
+		OLINUXINO_CONFIG(EMMC, GBYTES(4), GBYTES(2), COM)
+	},
+	{
+		OLINUXINO_BOARD(8857, "A64-OLinuXino-1G")
+		OLINUXINO_CONFIG(NONE, -1, GBYTES(1), COM)
+	},
+	{
+		OLINUXINO_BOARD(9849, "A64-OLinuXino-1Ge16GW")
+		OLINUXINO_CONFIG(EMMC, GBYTES(16), GBYTES(2), COM)
+	},
+#endif
+	{
+		.id = 0,
+	}
+};
